### PR TITLE
PDE-Refiner

### PR DESCRIPTION
**Pull request by phlippe**
This PR adds the code for [PDE-Refiner](https://arxiv.org/abs/2308.05732). The main changes are:
* Adding 1D datasets, specifically Kuramoto-Sivashinsky (dataset has to be uploaded to huggingface space);
* Adding a 1D conditional U-Net;
* Adding PDE-Refiner as separate model besides standard PDE surrogates and conditional ones;
* Adding Pearson correlation as additional rollout criteria;
* Adding Exponential Moving Average (EMA) of weights for better validation performance tracking.

PDE-Refiner has been additionally added under `research` in the docs and the training instructions to `train.md`.

### CODE REVIEW

File: pdearena/data/oned/datapipes/kuramotosivashinsky1d.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+                        f"WARNING: dt is not in the expected range (min {data['dx'].min()}, max {data['dx'].max()}, mean {data['dx'].mean()}) - scaling may be incorrect."
+                    )
+
+                if "v" in data:
+                    if data["v"].min() >= 0.5 and data["v"].max() <= 1.5:
+                        data["v"] = (data["v"] - 0.5) * 100.0
+                    else:
+                        print(
+                            f"WARNING: v is not in the expected range (min {data['v'].min()}, max {data['v'].max()}, mean {data['v'].mean()}) - scaling may be incorrect."
+                        )
```
Comment (rejuvyesh) [inline pdearena/data/oned/datapipes/kuramotosivashinsky1d.py:99]: Shouldn't these have happened in data generation instead?

File: pdearena/models/pderefiner.py [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+
+    def on_train_batch_end(self, *args, **kwargs):
+        self.ema.update()
+
+    def on_validation_start(self):
+        self.apply_ema()
+
+    def on_validation_end(self):
+        self.remove_ema()
+
```
Comment (rejuvyesh) [inline pdearena/models/pderefiner.py:410]: Should also be done for `on_test_start` and `on_test_end`?

File: scripts/pderefiner_train.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        PDERefiner,
+        PDEDataModule,
+        seed_everything_default=42,
+        save_config_overwrite=True,
+        run=False,
+        parser_kwargs={"parser_mode": "omegaconf"},
+    )
+    if cli.trainer.default_root_dir is None:
+        logger.warning("No default root dir set, using: ")
+        cli.trainer.default_root_dir = os.environ.get("AMLT_OUTPUT_DIR", "./outputs")
```
Comment (rejuvyesh) [inline scripts/pderefiner_train.py:40]:
```suggestion
        cli.trainer.default_root_dir = os.environ.get("PDEARENA_OUTPUT_DIR", "./outputs")
```

**Review by rejuvyesh — CHANGES_REQUESTED** [review]
Can you also add `diffusers` required version to setup.py as it's not an easy dependency with frequent breaking updates?

File: scripts/pderefiner_train.py (hunk 3 above)
Comment (phlippe) [reply scripts/pderefiner_train.py]: Thanks! Done.

File: pdearena/models/pderefiner.py (hunk 2 above)
Comment (phlippe) [reply pdearena/models/pderefiner.py:410]: Good point. Usually we test after having finished the training with a learning rate cooldown, which makes the EMA basically identical to the original weights. But to be consistent, let's also go with applying the EMA during testing.

File: configs/kuramotosivashinsky1d.yaml [hunk 4]
```diff
@@ -0,0 +1,126 @@
+seed_everything: 42
+trainer:
+  default_root_dir: ${oc.env:AMLT_OUTPUT_DIR,outputs}
```
[the remark this answers is not in the document]
Comment (phlippe) [reply configs/kuramotosivashinsky1d.yaml]: Thanks! Done.

One thing to note: the other configs seem to also use `AMLT_OUTPUT_DIR`. Maybe we should change it for all of them to be consistent?

File: pdearena/data/oned/datapipes/common.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+                dpipe,
+                pde.n_scalar_components,
+                pde.n_vector_components,
+                pde.n_scalar_components,
+                pde.n_vector_components,
+                pde.trajlen,
+                time_history,
+                time_future,
+                time_gap,
+            )
```
[the remark this answers is not in the document]
Comment (phlippe) [reply pdearena/data/oned/datapipes/common.py:94]: Yes, let's add that (probably in the refactor). In the 2D case, it also seems to require `onestep` to always be true for eval, right?

File: pdearena/data/oned/datapipes/kuramotosivashinsky1d.py (hunk 1 above)
Comment (phlippe) [reply pdearena/data/oned/datapipes/kuramotosivashinsky1d.py:99]: Hm, I had preferred to do them in the data loading since `dt` and `dx` have units that are easier to relate to (i.e. seconds / meters). Rescaling them already in the data removes this relation, especially if someone wants to read them out for other purposes. An alternative would be to add extra keys in the hdf5 file like `scaled_dt` and `scaled_dx` which contain then this calculation (`dt` and `dx` are quite cheap to store). Would that be preferable?

**Comment by phlippe** [comment]
> Can you also add `diffusers` required version to setup.py as it's not an easy dependency with frequent breaking updates?

Thanks for the quick review! Added the version that I am currently using (`0.17.1`).

File: pdearena/data/oned/datapipes/kuramotosivashinsky1d.py (hunk 1 above)
Comment (rejuvyesh) [reply pdearena/data/oned/datapipes/kuramotosivashinsky1d.py:99]: Ah, it's alright then. Maybe should add a short note because without context it feels like we are changing the underlying data behind the phenomenon itself to emphasize certain properties.

File: pdearena/data/oned/datapipes/common.py [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+            # Length of trajectory
+            time_resolution = min(self.trajlen, u.shape[0])
+            # Max number of previous points solver can eat
+            reduced_time_resolution = time_resolution - self.time_history
+            # Number of future points to predict
+            max_start_time = reduced_time_resolution - self.time_future - self.time_gap
+
+            # Choose initial random time point at the PDE solution manifold
+            start_time = random.choices([t for t in range(max_start_time + 1)], k=1)
+            data, targets = datautils.create_data2D(
```
[the remark this answers is not in the document]
Comment (rejuvyesh) [reply pdearena/data/oned/datapipes/common.py:185]: Hmm, we did have 2D `cond` dataset loading as well? But yeah would be better to modify the current implementation and skipping when it's `None`. Can help fix if it breaks the current 2D pipeline but should be easy enough to fix.

File: configs/kuramotosivashinsky1d.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (rejuvyesh) [reply configs/kuramotosivashinsky1d.yaml]: Yep! AMLT is microsoft internal thing, so best to avoid.

File: pdearena/data/oned/datapipes/common.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (phlippe) [reply pdearena/data/oned/datapipes/common.py]: Refactored in 2e697b15e5ec38d4b44ca8ff1ccdc1d08786be3e

File: pdearena/data/oned/datapipes/kuramotosivashinsky1d.py (hunk 1 above)
Comment (phlippe) [reply pdearena/data/oned/datapipes/kuramotosivashinsky1d.py]: Added a clarification in c96734ba2fd1bd07387d40f603d1e4110f3c7b8b

File: configs/kuramotosivashinsky1d.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (phlippe) [reply configs/kuramotosivashinsky1d.yaml]: Changed the remaining occurrences of `AMLT_OUTPUT_DIR` to `PDEARENA_OUTPUT_DIR` in dc91f8635f3cfac5bee21d618bfbd4c06f571695

**Comment by phlippe** [comment]
I tried to address the remaining points, let me know if there are any other open points :)

I'm currently running a final experiment to verify the results in the code base here. Previous experiments already showed it working as expected in most metrics, only a fix for the rollout was missing and added in a commit now.

File: docs/research.md [hunk 7]
```diff
@@ -7,6 +7,30 @@ Following is a list of research papers that have been published using the PDEAre
 If you have used PDEArena in your research, and would like it listed here, please add your paper to [this file](https://github.com/microsoft/pdearena/blob/main/docs/research.yml) by sending a pull request to the [PDEArena repository](https://github.com/microsoft/pdearena).
 
 
+<figure markdown>
```
Comment (rejuvyesh) [inline docs/research.md:10]: This is auto generated based on https://github.com/microsoft/pdearena/blob/main/docs/research.yml. So better to modify that instead.

Comment (phlippe) [reply docs/research.md]: Ah, thanks! Somehow had missed that file. Moved the paper addition to the yaml file now.

**Review by rejuvyesh — APPROVED** [review]

**Review by rejuvyesh — COMMENTED** [review]
Can you give me access to your branch? OR run `pre-commit run` to automatically format the files?

**Comment by phlippe** [comment]
> Can you give me access to your branch? OR run `pre-commit run` to automatically format the files?

The `pre-commit run` currently fails for me locally, but seems to be an installation problem of the pre-commit components. I added you as a collaborator to my fork, hopefully that gives you access to the branch.

**Comment by rejuvyesh** [comment]
Tests pass for me locally. Would be great if you could add a test for 1D data loading and 1D Unet. But not a blocker and will merge soon.

**Comment by phlippe** [comment]
Sure, let me take a look and implement some tests now.

**Comment by phlippe** [comment]
I added tests for the 1D data loading and the conditional 1D U-Nets (pass locally). I noticed that the 2D data loading test failed now because the conditioning was not None and the refactored datapipes now return the conditioning during training. Not sure if there is some training code that needs to be adjusted accordingly as well.

**Comment by rejuvyesh** [comment]
```
FAILED tests/test_model_registry.py::test_named_models[MaxwellFNO3D-96-8] - AssertionError
FAILED tests/test_model_registry.py::test_named_models[MaxwellCFNO3D-32-8] - AssertionError
```

@brandstetter-johannes can you take a look, when you get the time? Maxwell3D / Clifford tests seem to be failing.

**Comment by rejuvyesh** [comment]
Remaining fixes can be separate PRs.

**Comment by scaomath** [comment]
@rejuvyesh Hey Jayesh, nice work on PDE-refiner. Sorry if this is somewhere in repo. I wonder where can I find the Kolmogorov flow ground truth data.

**Comment by rejuvyesh** [comment]
@scaomath You mean https://microsoft.github.io/pdearena/datadownload/#kuramoto-sivashinsky-1d ?

**Comment by brandstetter-johannes** [comment]
@scaomath the Kolmogorov flow data is taken from this paper: https://arxiv.org/abs/2302.08105. We are happy to put it on huggingface as well if Phillip finds the time

**Comment by scaomath** [comment]
@brandstetter-johannes @rejuvyesh Thanks a lot. That's very interesting. I wonder if the DNS baseline (FVM on a staggered grid on 2048^2 mesh) time is obtained from running the data gen code given in https://arxiv.org/abs/2302.08105 ?

**Comment by brandstetter-johannes** [comment]
Yes it is, we rerun the whole code

**Comment by scaomath** [comment]
Nice thanks for the clarification. BTW is this repo welcoming PR? or just your group at MS.
